### ISSUE TITLE
A Hard-to-Describe Numerical Error in nn.Linear for Tensors with dim>2

### ISSUE THREAD

**Issue by jachiam** [pytorch/pytorch#91598]
### 🐛 Describe the bug

```
import torch
import torch.nn as nn

torch.manual_seed(0)

def test_linear():
    X1 = torch.rand(5,8)
    X2 = torch.rand(5,8)
    X = torch.cat([X1, X2], dim=0) # 10, 8

    lin = nn.Linear(8,11, bias=False)

    Y1 = lin(X1) # 5, 11
    Y2 = lin(X2) # 5, 11
    Y = lin(X) # 10, 11

    Y_ = torch.cat([Y1, Y2], dim=0)

    abs_diff = torch.abs(Y - Y_)
    print(abs_diff.mean(), abs_diff.min(), abs_diff.max())

    assert torch.all(Y == Y_)


def test_linear_multidim():
    X1 = torch.rand(2,5,8)
    X2 = torch.rand(2,5,8)
    X = torch.cat([X1, X2], dim=1) # 2, 10, 8

    lin = nn.Linear(8,11, bias=False)

    Y1 = lin(X1) # 2, 5, 11
    Y2 = lin(X2) # 2, 5, 11
    Y = lin(X) # 2, 10, 11

    Y_ = torch.cat([Y1, Y2], dim=1)

    abs_diff = torch.abs(Y - Y_)
    print(abs_diff.mean(), abs_diff.min(), abs_diff.max())

    assert torch.all(Y == Y_)
```

The first test passes, the second test fails.

The gist of the issue: the tests are checking that if you have X = cat([X1, X2]), you can apply a linear layer to X1, then to X2, and concatenate the results to get the output from the linear layer applied to X. It works when X1 and X2 are matrices, but not when they are higher order tensors. The errors are small (~1e-7), but that's still quite surprising to me. That is to say - for some reason, whether you do unrelated parts of the matmul separately or together appears relevant somehow.

This is confusing - did I miss something really obvious? Is this expected behavior (just normal numerical error somehow)? Did I write a bugged test?

### Versions

```
PyTorch version: 1.12.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA RTX A2000 Laptop GPU
Nvidia driver version: 520.61.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] torch==1.12.0+cu116
[pip3] torchaudio==0.12.0+cu116
[pip3] torchvision==0.13.0+cu116
[conda] numpy                     1.23.1                   pypi_0    pypi
[conda] torch                     1.12.0+cu116             pypi_0    pypi
[conda] torchaudio                0.12.0+cu116             pypi_0    pypi
[conda] torchvision               0.13.0+cu116             pypi_0    pypi
```

**Comment by jachiam** [comment]
Ah, nevermind, this is actually described in the docs.

https://pytorch.org/docs/stable/notes/numerical_accuracy.html
